File: wrappers/java/foundation/src/test/java/com/virgilsecurity/crypto/foundation/RecipientCipherTest.java
package com.virgilsecurity.crypto.foundation;

import static org.junit.Assert.assertArrayEquals;
import static org.junit.Assert.assertNotNull;

import java.nio.charset.StandardCharsets;

import org.apache.commons.lang.ArrayUtils;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;

public class RecipientCipherTest extends SampleBasedTest {

	private static final byte[] CUSTOM_PARAM_SIGNATURE = "VIRGIL-DATA-SIGNATURE".getBytes(StandardCharsets.UTF_8);
	private static final byte[] CUSTOM_PARAM_SIGNER_ID = "VIRGIL-DATA-SIGNER-ID".getBytes(StandardCharsets.UTF_8);

	private RecipientCipher recipientCipher;

	@Before
	public void setup() {
		this.recipientCipher = new RecipientCipher();
	}

	@After
	public void tearDown() {
		this.recipientCipher.close();
	}

	@Test
	public void encrypt_decrypt__with_ed25519() {
		byte[] data = getBytes("data");
		byte[] recipientId = new byte[] { 0x01, 0x02, 0x03 };

		try (Ed25519 ed = new Ed25519()) {
			ed.setupDefaults();

			try (RawPrivateKey privateKey = (RawPrivateKey) ed.generateKey()) {

				// Encrypt
				this.recipientCipher.addKeyRecipient(recipientId, privateKey.extractPublicKey());
				this.recipientCipher.customParams().addData(CUSTOM_PARAM_SIGNER_ID, CUSTOM_PARAM_SIGNER_ID);
				this.recipientCipher.customParams().addData(CUSTOM_PARAM_SIGNATURE, CUSTOM_PARAM_SIGNATURE);

				this.recipientCipher.startEncryption();

				byte[] messageInfo = this.recipientCipher.packMessageInfo();
				assertNotNull(messageInfo);

				byte[] processEncryptionData = this.recipientCipher.processEncryption(data);
				assertNotNull(processEncryptionData);

				byte[] finishEncryptionData = this.recipientCipher.finishEncryption();
				assertNotNull(finishEncryptionData);

				byte[] encryptedData = ArrayUtils.addAll(messageInfo,
						ArrayUtils.addAll(processEncryptionData, finishEncryptionData));

				// Decrypt
				try (RecipientCipher cipher = new RecipientCipher()) {
					cipher.startDecryptionWithKey(recipientId, privateKey, new byte[0]);

					byte[] processDecryptionData = cipher.processDecryption(encryptedData);
					assertNotNull(processDecryptionData);

					byte[] finishDecryptionData = cipher.finishDecryption();
					assertNotNull(finishDecryptionData);

					byte[] decryptedData = ArrayUtils.addAll(processDecryptionData, finishDecryptionData);
					assertArrayEquals(data, decryptedData);
				}
			}
		}
	}

	@Test
	public void encrypt_decrypt__with_rsa() {
		byte[] data = getBytes("data");
		byte[] recipientId = new byte[] { 0x01, 0x02, 0x03 };

		try (Rsa rsa = new Rsa()) {
			rsa.setupDefaults();

			try (RsaPrivateKey privateKey = (RsaPrivateKey) rsa.generateKey(2048)) {

				// Encrypt
				this.recipientCipher.addKeyRecipient(recipientId, privateKey.extractPublicKey());
				this.recipientCipher.customParams().addData(CUSTOM_PARAM_SIGNER_ID, CUSTOM_PARAM_SIGNER_ID);
				this.recipientCipher.customParams().addData(CUSTOM_PARAM_SIGNATURE, CUSTOM_PARAM_SIGNATURE);

				this.recipientCipher.startEncryption();

				byte[] messageInfo = this.recipientCipher.packMessageInfo();
				assertNotNull(messageInfo);

				byte[] processEncryptionData = this.recipientCipher.processEncryption(data);
				assertNotNull(processEncryptionData);

				byte[] finishEncryptionData = this.recipientCipher.finishEncryption();
				assertNotNull(finishEncryptionData);

				byte[] encryptedData = ArrayUtils.addAll(messageInfo,
						ArrayUtils.addAll(processEncryptionData, finishEncryptionData));

				// Decrypt
				try (RecipientCipher cipher = new RecipientCipher()) {
					cipher.startDecryptionWithKey(recipientId, privateKey, new byte[0]);

					byte[] processDecryptionData = cipher.processDecryption(encryptedData);
					assertNotNull(processDecryptionData);

					byte[] finishDecryptionData = cipher.finishDecryption();
					assertNotNull(finishDecryptionData);

					byte[] decryptedData = ArrayUtils.addAll(processDecryptionData, finishDecryptionData);
					assertArrayEquals(data, decryptedData);
				}
			}
		}
	}

}

File: wrappers/wasm/tests/foundation/RecipientCipher.test.js
const initFoundation = require('../../dist/foundation/node.cjs');
const { hexToUint8Array } = require('../utils');

describe('RecipientCipher', () => {
  let foundation;
  let recipientCipher;

  beforeEach(async () => {
    foundation = await initFoundation();
    recipientCipher = new foundation.RecipientCipher();
  });

  test('encrypt / decrypt with ed25519 key pair', () => {
    const ed25519 = new foundation.Ed25519();
    ed25519.setupDefaults();
    const privateKey = ed25519.generateKey();
    const publicKey = privateKey.getPublicKey();
    const recipientId = hexToUint8Array('726563697069656e74');
    recipientCipher.addKeyRecipient(recipientId, publicKey);
    recipientCipher.startEncryption();
    const data = hexToUint8Array('64617461');
    const messageInfo = recipientCipher.packMessageInfo();
    const processEncryption = recipientCipher.processEncryption(data);
    const finishEncryption = recipientCipher.finishEncryption();
    const encryptedMessage = new Uint8Array(
      messageInfo.length + processEncryption.length + finishEncryption.length,
    );
    encryptedMessage.set(messageInfo);
    encryptedMessage.set(processEncryption, messageInfo.length);
    encryptedMessage.set(finishEncryption, messageInfo.length + processEncryption.length);
    recipientCipher.startDecryptionWithKey(recipientId, privateKey, new Uint8Array());
    const processDecryption = recipientCipher.processDecryption(encryptedMessage);
    const finishDecryption = recipientCipher.finishDecryption();
    const result = new Uint8Array(processDecryption.length + finishDecryption.length);
    result.set(processDecryption);
    result.set(finishDecryption, processDecryption.length);
    expect(result.toString()).toBe(data.toString());
  });
});
